homeassistant/components/pvpc_hourly_pricing/sensor.py

```diff
@@ -0,0 +1,195 @@
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._pvpc_data.state
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._pvpc_data.state_available
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return self._pvpc_data.attributes
+
+    async def async_update(self, *_args):
+        """Update the sensor state."""
+        if not self._init_done:
+            # abort until added_to_hass is finished
+            return
+
+        now = dt_util.utcnow()
+        if self._pvpc_data.process_state_and_attributes(now):
+            await self.async_update_ha_state()
+        else:
+            # If no prices present, download and schedule a future state update
+            self._pvpc_data.state_available = False
+            self.async_schedule_update_ha_state()
+
+            if self._pvpc_data.source_available:
+                _LOGGER.debug(
+                    "[%s]: Downloading prices as there are no valid ones",
+                    self.entity_id,
+                )
+                async_track_point_in_time(
+                    self.hass,
+                    self.async_update,
+                    now + timedelta(seconds=self._pvpc_data.timeout),
+                )
+            await self.async_update_prices(now)
+
+    async def async_update_prices(self, now):
+        """Update electricity prices from the ESIOS API."""
+        prices = await self._pvpc_data.async_update_prices(now)
+        if not prices and self._pvpc_data.source_available:
+            self._num_retries += 1
+            if self._num_retries > 2:
+                _LOGGER.warning(
+                    "Repeated bad data update, mark component as unavailable source"
+                )
+                self._pvpc_data.source_available = False
+                return
+
+            retry_delay = 3 * self._pvpc_data.timeout
+            _LOGGER.debug(
+                "Bad update[retry:%d], will try again in %d s",
+                self._num_retries,
+                retry_delay,
+            )
+            async_track_point_in_time(
+                self.hass,
+                self.async_update_prices,
+                dt_util.now() + timedelta(seconds=retry_delay),
+            )
+            return
+
+        if not prices:
+            _LOGGER.debug(
+                "Data source unavailable since %s",
+                self.hass.states.get(self.entity_id).last_changed,
+            )
+            return
+
+        self._num_retries = 0
+        if not self._pvpc_data.source_available:
+            self._pvpc_data.source_available = True
+            _LOGGER.warning(
+                "Component has recovered data access. Was unavailable since %s",
+                self.hass.states.get(self.entity_id).last_changed,
```

Comment:
Can you call
`self.last_changed` instead?

Reply:
🤯